src/kubernetes_cluster/spec/controller/controller_runtime.rs

```diff
@@ -48,52 +49,86 @@ pub open spec fn continue_reconcile<K: ResourceView, R: Reconciler<K>>() -> Cont
+            } else if reconcile_state.lib_response.is_Some() {
+                Option::Some(ResponseView::ExternalResponse(reconcile_state.lib_response.get_Some_0()))
             } else {
                 Option::None
             };
-            let cr_key = input.scheduled_cr_key.get_Some_0();
-            let reconcile_state = s.ongoing_reconciles[cr_key];
-
             let (local_state_prime, req_o) = R::reconcile_core(reconcile_state.triggering_cr, resp_o, reconcile_state.local_state);
+            if req_o.is_Some() {
+                match req_o.get_Some_0() {
+                    RequestView::KRequest(req) => {
+                        let (rest_id_allocator_prime, pending_req_msg) 
+                            = (input.rest_id_allocator.allocate().0, Option::Some(controller_req_msg(req, input.rest_id_allocator.allocate().1)));
+                        let reconcile_state_prime = OngoingReconcile {
+                            pending_req_msg: pending_req_msg,
+                            lib_response: Option::None,
+                            local_state: local_state_prime,
+                            ..reconcile_state
+                        };
+                        let s_prime = ControllerState {
+                            ongoing_reconciles: s.ongoing_reconciles.insert(cr_key, reconcile_state_prime),
+                            ..s
+                        };
+                        (s_prime, (Multiset::singleton(pending_req_msg.get_Some_0()), rest_id_allocator_prime))
+                    },
+                    RequestView::ExternalRequest(req) => {
+                        // Call the external library first.
 
-            let (rest_id_allocator_prime, pending_req_msg) = if req_o.is_Some() && req_o.get_Some_0().is_KRequest() {
-                (
-                    input.rest_id_allocator.allocate().0,
-                    Option::Some(controller_req_msg(req_o.get_Some_0().get_KRequest_0(), input.rest_id_allocator.allocate().1))
-                )
-            } else {
-                (input.rest_id_allocator, Option::None)
-            };
-
-            let reconcile_state_prime = OngoingReconcile {
-                pending_req_msg: pending_req_msg,
-                local_state: local_state_prime,
-                ..reconcile_state
-            };
-            let s_prime = ControllerState {
-                ongoing_reconciles: s.ongoing_reconciles.insert(cr_key, reconcile_state_prime),
-                ..s
-            };
-            let send = if pending_req_msg.is_Some() {
-                Multiset::singleton(pending_req_msg.get_Some_0())
+                        // Then update the state.
+                        let reconcile_state_prime = OngoingReconcile {
+                            pending_req_msg: Option::None,
+                            lib_response: Option::None, // TODO: update this field after add the library as a generic
```

Comment:
Do you want to leave it for the next PR of finish it in this one?

Reply:
Either work. But I must first let the verification go through before adding one more generic to the `Reconciler` trait.

Reply:
> But I must first let the verification go through before adding one more generic to the Reconciler trait.

Agree